Store: Add stats widget to the store dashboard

justinshreve opened the pull request:
This PR implements a stats widget on the dashboard. See p90Yrv-mi-p2.

It is wrapped behind a new feature flag, enabled in development and wpcalypso. 

To Test:
* Run `npm run test-client client/extensions/woocommerce` and make sure all tests pass.
* Go to `http://calypso.localhost:3000/store/:site` on a site with some orders. Test switching the unit (day, week, month). Refresh and make sure the selection persists.
* Verify the different stats/lists load and display correctly.
* If you don’t have a test site with a lot of data, you can force the widget to display another site as an A12.
	* Find a live site by going to Live Tracks view in MC and then search for `calypso_woocommerce_order_received`.  Find an event and expand it. Copy the blogid.
	* Open `client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js`
	* Replace `const site = getSelectedSiteWithFallback( state );` with something like `const site = { ID: BLOGID, slug: '', name: 'Test Site' };`.
	* Refresh

<img width="844" alt="screen shot 2018-03-15 at 1 26 56 am" src="https://user-images.githubusercontent.com/689165/37486137-d09c248a-284a-11e8-8aac-29b4b485aa5a.png">

Apologies for the PR size. About 200 lines are tests, and 130 CSS. I could split out the utils and tests (~300 lines) — but I figured it would be easier to review with context.

justinshreve commented:
Kelly mentioned on the thread:

> I thought of something… We’ll need a ‘zero state’ that doesn’t look/feel super sad if you have no orders, etc in the last week.

This is not implemented yet. I’m happy to do that in a follow-up PR before launching this. What should it look like? @kellychoffman @jameskoster 

This is what it currently looks like with no orders:

<img width="841" alt="screen shot 2018-03-15 at 1 03 15 am" src="https://user-images.githubusercontent.com/689165/37486220-17833794-284b-11e8-86b2-e32176d0fa50.png">

a8ci18n commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js:
```diff
@@ -0,0 +1,363 @@
+		return (
+			<Fragment>
+				<span>
+					{ translate( '%(siteName)s in the last {{timePeriodSelector/}}', {
```
:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

timmyc commented:
This. Is. Awesome.

The "sad panda" state - in other explorations of store system dashboards - it seems this isn't something that is typically handled very well. I agree that we could circle back on it and try some different approaches, but really having the numbers there, even when zero/sad - are pretty key pieces of information for a store operator. Perhaps showing the zeros is a motivator on its own.

Other ideas though:

* Maybe a prompt to write a post about a new product if one was added recently
* Plug publicize if no social connections exist on the site
* Provide insights to referrers that have a higher conversion rate so efforts could be focused there

kellychoffman commented:
> This is what it currently looks like with no orders:

I actually think this is okay for now, until we have something designed up.

kellychoffman commented:
One thing I realized when testing this is that we only show the mamma dashboard when there are new orders that need fulfilling. When there are no orders, we revert back to the fledgling dashboard, which was only intended to show until a store gets their first oder, and then never display again. I bring this up now, because I was expecting to see the stats widget even whether I had a new order or not. This is probably outside this PR, but wanted to mention it.

timmyc commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/stat.js:
```diff
@@ -0,0 +1,117 @@
+			case 'currency':
+				return formatCurrency( value, data.currency );
+			case 'percent':
+				return value + '%';
```
I have no idea but, do we need to have this be a translated string? Maybe in RTL this could be backwards?

kellychoffman commented:
Okay real review!

1. Remove Reports widget as its no longer needed. 
2. Move new stats widget underneath the New Order and Reviews widget so those are closer to the Hi, "Kelly Hoffman. You have new orders 🎉 header"
3. "View full stats" text link might be better (more visible) as a button.
4. Left align the text in the Day, Week, Month dropdown

timmyc commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/stat.js:
```diff
@@ -0,0 +1,117 @@
+	render() {
+		const { data, site, date, attribute, label, requesting } = this.props;
+
+		if ( requesting || ! site.ID || ! data.length ) {
```
should we add `! data` here too?

timmyc commented on client/extensions/woocommerce/app/store-stats/utils.js:
```diff
@@ -145,3 +173,59 @@ export function getDelta( deltas, selectedDate, stat ) {
+ * @return {object} - Object containing data from calculateDelta
+ */
+export function getDeltaFromData( data, selectedDate, stat, unit ) {
+	if ( data.length < 3 ) {
```
Yeah seems here we might be able to add a `! data && Array.isArray( data )` to simplify some of the usage of this in the components.

timmyc commented:
SO COOL!!!!

@kellychoffman beat me to a couple items that I noticed to, but will +1 on:

```
Move new stats widget underneath the New Order and Reviews widget so those are closer to the Hi, "Kelly Hoffman. You have new orders 🎉 header"
"View full stats" text link might be better (more visible) as a button.
```

Few other items I noticed:
![stats-1](https://user-images.githubusercontent.com/22080/37489199-1d0cf296-2854-11e8-9f69-1a044fbd8227.png)

Referrers Stat: The "More" link is being shown even though all are already being shown. Also the more link goes to the "main" store stats page right now... I guess until the referrer widget launches that will be a little funky.

More Links: As for the more links themselves - wondering if we need to make them a bit more distinct too. In site stats modules - these links are shown like:

![stats-view-all](https://user-images.githubusercontent.com/22080/37489352-912cd6f0-2854-11e8-9819-4524f66981be.png)

In that awkward screen size before the mobile breakpoint, things get a little odd in spots:

![stats-2](https://user-images.githubusercontent.com/22080/37489385-a75d4860-2854-11e8-815d-7e90a4ad7a9a.png)

Given the size of this PR, and that we have a feature flag - I'd say lets get this in, and do some small cleanup tasks around those items after because they are very edge-case. I also highlighted the View Full stats button as Kelly pointed out too.

psealock commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js:
```diff
@@ -0,0 +1,363 @@
+		const { site, translate, unit } = this.props;
+
+		const options = [
+			{ value: 'day', label: 'day' },
```
add translations to this block

psealock commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js:
```diff
@@ -0,0 +1,363 @@
+		const { orderQuery, topEarnersQuery, referrerQuery, visitorQuery } = queries;
+		return (
+			<Fragment>
+				<QueryPreferences />
```
I haven't taken this for a spin yet, but does `site` always exist? I'm not familiar with `getSelectedSiteWithFallback` just yet

justinshreve replied:
> I haven't taken this for a spin yet, but does site always exist? 

It should, because the app wrapper [doesn't render anything until site is available](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/app/index.js#L86). However, I can add a check here just to be safe.

justinshreve replied:
> Quite a few new queries - but fortunately most are against the wpcom api.

Yeah. All the stats ones are against .com and the synced data. The preferences one is also a .com endpoint. So no extra Jetpack/Proxy calls.

psealock commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/queries.js:
```diff
@@ -0,0 +1,48 @@
+import { getUnitPeriod, getStartDate } from 'woocommerce/app/store-stats/utils';
+import { UNITS, dashboardListLimit } from 'woocommerce/app/store-stats/constants';
+
+export const getQueries = unit => {
```
Nice! If you can make `getQueries` a function of `unit` and `dateString` and maybe move this file to `/store-stats`, I could reuse it for the other dashboard, keeping this logic all in one place.

Usage would be like:
```js
const{ orderQuery } = getQueries( unit, moment().format( 'YYYY-MM-DD' ) );
```

justinshreve replied:
I haven't done this yet but we can in a follow-up PR. Will this end up being useful for you? i.e. my referrers query only pulls 1 unit's worth of data. Wouldn't you need to pull more than that for some of your displays?

psealock replied:
It would be useful. But now I see what you mean. The difference is `quantity: 1` on the `referrerQuery`. Hmm...

Could we set it to `quantity: UNITS[ unit ].quantity` and override somehow if desired?

psealock commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/stat.js:
```diff
@@ -0,0 +1,117 @@
+		);
+	};
+
+	renderValue = ( value, data ) => {
```
Could you add a `percent` switch option to this existing function instead?

https://github.com/Automattic/wp-calypso/blob/8c16b16107342a659d17c29e90254218540d8d52/client/extensions/woocommerce/app/store-stats/utils.js#L124-L134

psealock commented on client/extensions/woocommerce/app/store-stats/store-stats-list/index.js:
```diff
@@ -47,10 +47,14 @@ const StoreStatsList = ( { data, values } ) => {
 StoreStatsList.propTypes = {
 	data: PropTypes.array.isRequired,
 	values: PropTypes.array.isRequired,
+	fetchedData: PropTypes.oneOfType( [ PropTypes.array, PropTypes.object ] ),
```
👍

ryelle commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js:
```diff
@@ -0,0 +1,363 @@
+	dateForDisplay = () => {
+		const { translate, unit } = this.props;
+
+		const localizedDate = moment( moment().format( 'YYYY-MM-DD' ) );
```
The `localize` HOC also gives you `moment`, so you can grab this from props just like `translate`

justinshreve commented:
@timmyc @kellychoffman @psealock  @ryelle Thanks for the PR feedback! I've addressed most of the feedback here. If things are looking good, let's try to get this merged since it's behind a feature flag.

These are the items I'm leaving for follow-up PRs:

* Update/make getQueries more general purpose [see this comment](https://github.com/Automattic/wp-calypso/pull/23344/files/aea4c294a06bba130e63ec5f21c0633e5da6fbbd#diff-5472e06e1f7f41ddfd522c2edc4fe996).
* [Mobile CSS Fixes](https://github.com/Automattic/wp-calypso/pull/23344#issuecomment-373512421).
* Adding some Track events around the button, unit drop down, and links.

Full changelog of my updates:

* Removed the reports widgets on both the fledging and full dashboard views.
* Moved the stats widget under the new order/pending review widgets.
* Make "View full stats" a button.
* Left aligned the text in the Day, Week, Month dropdown.
* Update and use `formatValue` instead with % support. Also add translation wrapper around display.
* Changed up ‘more’ link display. Renamed, and products list now links to the products list, and the referrer link will only show up when the referrers page config flag is true.
* Fixed some translations/localization stuff (including using moment from the HOC).
* Some general code cleaning. Strengthen some checks. Fix delta function returns, etc.

a8ci18n commented on client/extensions/woocommerce/app/dashboard/widgets/stats-widget/index.js:
```diff
@@ -0,0 +1,366 @@
+		return (
+			<Fragment>
+				<span>
+					{ translate( '%(siteName)s in the last {{timePeriod/}}', {
```
:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/extensions/woocommerce/app/store-stats/utils.js:
```diff
@@ -127,6 +155,8 @@ export function formatValue( value, format, code ) {
 		case 'number':
 			return Math.round( value * 100 ) / 100;
+		case 'percent':
+			return translate( '%(percentage)s%% ', { args: { percentage: value }, context: 'percent' } );
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( '%(percentage)s%% completed' )` *ES Score: 16*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

greenafrican commented on client/extensions/woocommerce/app/store-stats/utils.js:
```diff
@@ -143,5 +173,60 @@ export function formatValue( value, format, code ) {
+		if ( visitorRow.visitors > 0 && orderData[ index ] ) {
+			return {
+				period: unitPeriod,
+				conversionRate: round( orders / visitorRow.visitors * 100, 2 ),
```
Quick one: are these uniques? If 1 customer makes 2 orders in the period, is that ( 2 orders / 1 visitor ) or ( 2 orders / 2 visitors ) OR ( 1 order / 1 visitor )?

justinshreve replied:
@greenafrican 

Visitors is unique. Orders is not. So that would be `( 2 orders / 1 visitor )`.

> then perhaps both stats should be uniques?

I don't think we have the data for unique orders as far as I can tell. Do you know of something?

When searching and doing some research, it looked like orders/actions divided by unique visitors works for calculating your conversion rate, as long as you are being consistent in your other calculations. We could have a help icon that shows what the calculation is?

--

It seems like our options are `orders / visitors` or `orders / unique visitors`

justinshreve replied:
I took a look to see if we could made add a new endpoint but yeah, I don't think we have the synced data to do unique orders at this time. See comment #comment-1059 on p90Yrv-Cl-p2. Anything we could use to do unique orders will no longer be synced. I also don't think we would get accurate uniques for guest orders this way. 

I would say we keep this calculation as is (with a little indication what the calculation is) and maybe long term we could add some tracking to Jetpack that helps us get a real uniques order count?
